// plugin-welcome/src/api/models/ControllersBookcourse.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API Patient
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface ControllersBookcourse
 */
export interface ControllersBookcourse {
    /**
     * 
     * @type {string}
     * @memberof ControllersBookcourse
     */
    booktime?: string;
    /**
     * 
     * @type {number}
     * @memberof ControllersBookcourse
     */
    course?: number;
    /**
     * 
     * @type {number}
     * @memberof ControllersBookcourse
     */
    employee?: number;
    /**
     * 
     * @type {number}
     * @memberof ControllersBookcourse
     */
    member?: number;
}

export function ControllersBookcourseFromJSON(json: any): ControllersBookcourse {
    return ControllersBookcourseFromJSONTyped(json, false);
}

export function ControllersBookcourseFromJSONTyped(json: any, ignoreDiscriminator: boolean): ControllersBookcourse {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'booktime': !exists(json, 'booktime') ? undefined : json['booktime'],
        'course': !exists(json, 'course') ? undefined : json['course'],
        'employee': !exists(json, 'employee') ? undefined : json['employee'],
        'member': !exists(json, 'member') ? undefined : json['member'],
    };
}

export function ControllersBookcourseToJSON(value?: ControllersBookcourse | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'booktime': value.booktime,
        'course': value.course,
        'employee': value.employee,
        'member': value.member,
    };
}
