// plugin-welcome/src/api/models/ControllersEmployee.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API Patient
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface ControllersEmployee
 */
export interface ControllersEmployee {
    /**
     * 
     * @type {number}
     * @memberof ControllersEmployee
     */
    age?: number;
    /**
     * 
     * @type {string}
     * @memberof ControllersEmployee
     */
    employeeaddress?: string;
    /**
     * 
     * @type {string}
     * @memberof ControllersEmployee
     */
    employeeid?: string;
    /**
     * 
     * @type {string}
     * @memberof ControllersEmployee
     */
    employeename?: string;
    /**
     * 
     * @type {string}
     * @memberof ControllersEmployee
     */
    idcardnumber?: string;
    /**
     * 
     * @type {number}
     * @memberof ControllersEmployee
     */
    position?: number;
    /**
     * 
     * @type {number}
     * @memberof ControllersEmployee
     */
    salary?: number;
}

export function ControllersEmployeeFromJSON(json: any): ControllersEmployee {
    return ControllersEmployeeFromJSONTyped(json, false);
}

export function ControllersEmployeeFromJSONTyped(json: any, ignoreDiscriminator: boolean): ControllersEmployee {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'age': !exists(json, 'age') ? undefined : json['age'],
        'employeeaddress': !exists(json, 'employeeaddress') ? undefined : json['employeeaddress'],
        'employeeid': !exists(json, 'employeeid') ? undefined : json['employeeid'],
        'employeename': !exists(json, 'employeename') ? undefined : json['employeename'],
        'idcardnumber': !exists(json, 'idcardnumber') ? undefined : json['idcardnumber'],
        'position': !exists(json, 'position') ? undefined : json['position'],
        'salary': !exists(json, 'salary') ? undefined : json['salary'],
    };
}

export function ControllersEmployeeToJSON(value?: ControllersEmployee | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'age': value.age,
        'employeeaddress': value.employeeaddress,
        'employeeid': value.employeeid,
        'employeename': value.employeename,
        'idcardnumber': value.idcardnumber,
        'position': value.position,
        'salary': value.salary,
    };
}
