// plugin-welcome/src/api/models/ControllersPromotiontype.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API Patient
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface ControllersPromotiontype
 */
export interface ControllersPromotiontype {
    /**
     * 
     * @type {string}
     * @memberof ControllersPromotiontype
     */
    type?: string;
}

export function ControllersPromotiontypeFromJSON(json: any): ControllersPromotiontype {
    return ControllersPromotiontypeFromJSONTyped(json, false);
}

export function ControllersPromotiontypeFromJSONTyped(json: any, ignoreDiscriminator: boolean): ControllersPromotiontype {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'type': !exists(json, 'type') ? undefined : json['type'],
    };
}

export function ControllersPromotiontypeToJSON(value?: ControllersPromotiontype | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'type': value.type,
    };
}
