// plugin-welcome/src/api/models/EntBookcourse.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API Patient
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EntBookcourseEdges,
    EntBookcourseEdgesFromJSON,
    EntBookcourseEdgesFromJSONTyped,
    EntBookcourseEdgesToJSON,
} from './';

/**
 * 
 * @export
 * @interface EntBookcourse
 */
export interface EntBookcourse {
    /**
     * BOOKTIME holds the value of the "BOOKTIME" field.
     * @type {string}
     * @memberof EntBookcourse
     */
    bOOKTIME?: string;
    /**
     * 
     * @type {EntBookcourseEdges}
     * @memberof EntBookcourse
     */
    edges?: EntBookcourseEdges;
    /**
     * ID of the ent.
     * @type {number}
     * @memberof EntBookcourse
     */
    id?: number;
}

export function EntBookcourseFromJSON(json: any): EntBookcourse {
    return EntBookcourseFromJSONTyped(json, false);
}

export function EntBookcourseFromJSONTyped(json: any, ignoreDiscriminator: boolean): EntBookcourse {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'bOOKTIME': !exists(json, 'BOOKTIME') ? undefined : json['BOOKTIME'],
        'edges': !exists(json, 'edges') ? undefined : EntBookcourseEdgesFromJSON(json['edges']),
        'id': !exists(json, 'id') ? undefined : json['id'],
    };
}

export function EntBookcourseToJSON(value?: EntBookcourse | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'BOOKTIME': value.bOOKTIME,
        'edges': EntBookcourseEdgesToJSON(value.edges),
        'id': value.id,
    };
}
