// plugin-welcome/src/api/models/EntEmployee.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API Patient
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EntEmployeeEdges,
    EntEmployeeEdgesFromJSON,
    EntEmployeeEdgesFromJSONTyped,
    EntEmployeeEdgesToJSON,
} from './';

/**
 * 
 * @export
 * @interface EntEmployee
 */
export interface EntEmployee {
    /**
     * EMPLOYEEADDRESS holds the value of the "EMPLOYEEADDRESS" field.
     * @type {string}
     * @memberof EntEmployee
     */
    eMPLOYEEADDRESS?: string;
    /**
     * EMPLOYEEID holds the value of the "EMPLOYEEID" field.
     * @type {string}
     * @memberof EntEmployee
     */
    eMPLOYEEID?: string;
    /**
     * EMPLOYEENAME holds the value of the "EMPLOYEENAME" field.
     * @type {string}
     * @memberof EntEmployee
     */
    eMPLOYEENAME?: string;
    /**
     * IDCARDNUMBER holds the value of the "IDCARDNUMBER" field.
     * @type {string}
     * @memberof EntEmployee
     */
    iDCARDNUMBER?: string;
    /**
     * 
     * @type {EntEmployeeEdges}
     * @memberof EntEmployee
     */
    edges?: EntEmployeeEdges;
    /**
     * ID of the ent.
     * @type {number}
     * @memberof EntEmployee
     */
    id?: number;
}

export function EntEmployeeFromJSON(json: any): EntEmployee {
    return EntEmployeeFromJSONTyped(json, false);
}

export function EntEmployeeFromJSONTyped(json: any, ignoreDiscriminator: boolean): EntEmployee {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'eMPLOYEEADDRESS': !exists(json, 'EMPLOYEEADDRESS') ? undefined : json['EMPLOYEEADDRESS'],
        'eMPLOYEEID': !exists(json, 'EMPLOYEEID') ? undefined : json['EMPLOYEEID'],
        'eMPLOYEENAME': !exists(json, 'EMPLOYEENAME') ? undefined : json['EMPLOYEENAME'],
        'iDCARDNUMBER': !exists(json, 'IDCARDNUMBER') ? undefined : json['IDCARDNUMBER'],
        'edges': !exists(json, 'edges') ? undefined : EntEmployeeEdgesFromJSON(json['edges']),
        'id': !exists(json, 'id') ? undefined : json['id'],
    };
}

export function EntEmployeeToJSON(value?: EntEmployee | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'EMPLOYEEADDRESS': value.eMPLOYEEADDRESS,
        'EMPLOYEEID': value.eMPLOYEEID,
        'EMPLOYEENAME': value.eMPLOYEENAME,
        'IDCARDNUMBER': value.iDCARDNUMBER,
        'edges': EntEmployeeEdgesToJSON(value.edges),
        'id': value.id,
    };
}
