// plugin-welcome/src/api/models/EntPosition.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API Patient
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EntPositionEdges,
    EntPositionEdgesFromJSON,
    EntPositionEdgesFromJSONTyped,
    EntPositionEdgesToJSON,
} from './';

/**
 * 
 * @export
 * @interface EntPosition
 */
export interface EntPosition {
    /**
     * POSITION holds the value of the "POSITION" field.
     * @type {string}
     * @memberof EntPosition
     */
    pOSITION?: string;
    /**
     * 
     * @type {EntPositionEdges}
     * @memberof EntPosition
     */
    edges?: EntPositionEdges;
    /**
     * ID of the ent.
     * @type {number}
     * @memberof EntPosition
     */
    id?: number;
}

export function EntPositionFromJSON(json: any): EntPosition {
    return EntPositionFromJSONTyped(json, false);
}

export function EntPositionFromJSONTyped(json: any, ignoreDiscriminator: boolean): EntPosition {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'pOSITION': !exists(json, 'POSITION') ? undefined : json['POSITION'],
        'edges': !exists(json, 'edges') ? undefined : EntPositionEdgesFromJSON(json['edges']),
        'id': !exists(json, 'id') ? undefined : json['id'],
    };
}

export function EntPositionToJSON(value?: EntPosition | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'POSITION': value.pOSITION,
        'edges': EntPositionEdgesToJSON(value.edges),
        'id': value.id,
    };
}
