// plugin-welcome/src/api/models/EntSalary.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API Patient
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EntSalaryEdges,
    EntSalaryEdgesFromJSON,
    EntSalaryEdgesFromJSONTyped,
    EntSalaryEdgesToJSON,
} from './';

/**
 * 
 * @export
 * @interface EntSalary
 */
export interface EntSalary {
    /**
     * SALARY holds the value of the "SALARY" field.
     * @type {number}
     * @memberof EntSalary
     */
    sALARY?: number;
    /**
     * 
     * @type {EntSalaryEdges}
     * @memberof EntSalary
     */
    edges?: EntSalaryEdges;
    /**
     * ID of the ent.
     * @type {number}
     * @memberof EntSalary
     */
    id?: number;
}

export function EntSalaryFromJSON(json: any): EntSalary {
    return EntSalaryFromJSONTyped(json, false);
}

export function EntSalaryFromJSONTyped(json: any, ignoreDiscriminator: boolean): EntSalary {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'sALARY': !exists(json, 'SALARY') ? undefined : json['SALARY'],
        'edges': !exists(json, 'edges') ? undefined : EntSalaryEdgesFromJSON(json['edges']),
        'id': !exists(json, 'id') ? undefined : json['id'],
    };
}

export function EntSalaryToJSON(value?: EntSalary | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'SALARY': value.sALARY,
        'edges': EntSalaryEdgesToJSON(value.edges),
        'id': value.id,
    };
}
