// plugin-welcome/src/api/models/EntZone.ts
/* tslint:disable */
/* eslint-disable */
/**
 * SUT SA Example API Patient
 * This is a sample server for SUT SE 2563
 *
 * The version of the OpenAPI document: 1.0
 * Contact: support@swagger.io
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EntZoneEdges,
    EntZoneEdgesFromJSON,
    EntZoneEdgesFromJSONTyped,
    EntZoneEdgesToJSON,
} from './';

/**
 * 
 * @export
 * @interface EntZone
 */
export interface EntZone {
    /**
     * EQUIPMENTZONE holds the value of the "EQUIPMENTZONE" field.
     * @type {string}
     * @memberof EntZone
     */
    eQUIPMENTZONE?: string;
    /**
     * 
     * @type {EntZoneEdges}
     * @memberof EntZone
     */
    edges?: EntZoneEdges;
    /**
     * ID of the ent.
     * @type {number}
     * @memberof EntZone
     */
    id?: number;
}

export function EntZoneFromJSON(json: any): EntZone {
    return EntZoneFromJSONTyped(json, false);
}

export function EntZoneFromJSONTyped(json: any, ignoreDiscriminator: boolean): EntZone {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'eQUIPMENTZONE': !exists(json, 'EQUIPMENTZONE') ? undefined : json['EQUIPMENTZONE'],
        'edges': !exists(json, 'edges') ? undefined : EntZoneEdgesFromJSON(json['edges']),
        'id': !exists(json, 'id') ? undefined : json['id'],
    };
}

export function EntZoneToJSON(value?: EntZone | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'EQUIPMENTZONE': value.eQUIPMENTZONE,
        'edges': EntZoneEdgesToJSON(value.edges),
        'id': value.id,
    };
}
