// @backstage/dev-utils/dist/index.esm.js
import { hot } from 'react-hot-loader/root';
import React from 'react';
import ReactDOM from 'react-dom';
import BookmarkIcon from '@material-ui/icons/Bookmark';
import { createApiFactory, alertApiRef, AlertApiForwarder, errorApiRef, ErrorAlerter, ErrorApiForwarder, oauthRequestApiRef, OAuthRequestManager, googleAuthApiRef, GoogleAuth, githubAuthApiRef, GithubAuth, gitlabAuthApiRef, GitlabAuth, createApp, SidebarItem, Sidebar, SidebarSpacer, ApiTestRegistry, AlertDisplay, OAuthRequestDialog, SidebarPage } from '@backstage/core';
import SentimentDissatisfiedIcon from '@material-ui/icons/SentimentDissatisfied';

const alertApiFactory = createApiFactory({
  implements: alertApiRef,
  deps: {},
  factory: () => new AlertApiForwarder()
});
const errorApiFactory = createApiFactory({
  implements: errorApiRef,
  deps: {alertApi: alertApiRef},
  factory: ({alertApi}) => new ErrorAlerter(alertApi, new ErrorApiForwarder())
});
const oauthRequestApiFactory = createApiFactory({
  implements: oauthRequestApiRef,
  deps: {},
  factory: () => new OAuthRequestManager()
});
const googleAuthApiFactory = createApiFactory({
  implements: googleAuthApiRef,
  deps: {oauthRequestApi: oauthRequestApiRef},
  factory: ({oauthRequestApi}) => GoogleAuth.create({
    apiOrigin: "http://localhost:7000",
    basePath: "/auth/",
    oauthRequestApi
  })
});
const githubAuthApiFactory = createApiFactory({
  implements: githubAuthApiRef,
  deps: {oauthRequestApi: oauthRequestApiRef},
  factory: ({oauthRequestApi}) => GithubAuth.create({
    apiOrigin: "http://localhost:7000",
    basePath: "/auth/",
    oauthRequestApi
  })
});
const gitlabAuthApiFactory = createApiFactory({
  implements: gitlabAuthApiRef,
  deps: {oauthRequestApi: oauthRequestApiRef},
  factory: ({oauthRequestApi}) => GitlabAuth.create({
    apiOrigin: "http://localhost:7000",
    basePath: "/auth/",
    oauthRequestApi
  })
});

var defaultApiFactories = /*#__PURE__*/Object.freeze({
  __proto__: null,
  alertApiFactory: alertApiFactory,
  errorApiFactory: errorApiFactory,
  oauthRequestApiFactory: oauthRequestApiFactory,
  googleAuthApiFactory: googleAuthApiFactory,
  githubAuthApiFactory: githubAuthApiFactory,
  gitlabAuthApiFactory: gitlabAuthApiFactory
});

class DevAppBuilder {
  constructor() {
    this.plugins = new Array();
    this.factories = new Array();
    this.rootChildren = new Array();
  }
  registerPlugin(...plugins) {
    this.plugins.push(...plugins);
    return this;
  }
  registerApiFactory(factory) {
    this.factories.push(factory);
    return this;
  }
  addRootChild(node) {
    this.rootChildren.push(node);
    return this;
  }
  build() {
    const app = createApp({
      apis: this.setupApiRegistry(this.factories),
      plugins: this.plugins
    });
    const AppProvider = app.getProvider();
    const AppRouter = app.getRouter();
    const AppRoutes = app.getRoutes();
    const sidebar = this.setupSidebar(this.plugins);
    const DevApp = () => {
      return /* @__PURE__ */ React.createElement(AppProvider, null, /* @__PURE__ */ React.createElement(AlertDisplay, null), /* @__PURE__ */ React.createElement(OAuthRequestDialog, null), this.rootChildren, /* @__PURE__ */ React.createElement(AppRouter, null, /* @__PURE__ */ React.createElement(SidebarPage, null, sidebar, /* @__PURE__ */ React.createElement(AppRoutes, null))));
    };
    return DevApp;
  }
  render() {
    const DevApp = hot(this.build());
    const paths = this.findPluginPaths(this.plugins);
    if (window.location.pathname === "/") {
      if (!paths.includes("/") && paths.length > 0) {
        window.location.pathname = paths[0];
      }
    }
    ReactDOM.render(/* @__PURE__ */ React.createElement(DevApp, null), document.getElementById("root"));
  }
  setupSidebar(plugins) {
    var _a;
    const sidebarItems = new Array();
    for (const plugin of plugins) {
      for (const output of plugin.output()) {
        switch (output.type) {
          case "legacy-route": {
            const {path} = output;
            sidebarItems.push(/* @__PURE__ */ React.createElement(SidebarItem, {
              key: path,
              to: path,
              text: path,
              icon: BookmarkIcon
            }));
            break;
          }
          case "route": {
            const {target} = output;
            sidebarItems.push(/* @__PURE__ */ React.createElement(SidebarItem, {
              key: target.path,
              to: target.path,
              text: target.title,
              icon: (_a = target.icon) != null ? _a : SentimentDissatisfiedIcon
            }));
            break;
          }
        }
      }
    }
    return /* @__PURE__ */ React.createElement(Sidebar, null, /* @__PURE__ */ React.createElement(SidebarSpacer, null), sidebarItems);
  }
  setupApiRegistry(providedFactories) {
    const providedApis = new Set(providedFactories.map((factory) => factory.implements));
    const defaultFactories = Object.values(defaultApiFactories).filter((factory) => !providedApis.has(factory.implements));
    const allFactories = [...defaultFactories, ...providedFactories];
    const registry = new ApiTestRegistry();
    for (const factory of allFactories) {
      registry.register(factory);
    }
    return registry;
  }
  findPluginPaths(plugins) {
    const paths = new Array();
    for (const plugin of plugins) {
      for (const output of plugin.output()) {
        if (output.type === "legacy-route") {
          paths.push(output.path);
        }
      }
    }
    return paths;
  }
}
function createDevApp() {
  return new DevAppBuilder();
}

export { createDevApp };
